XXX-asyncwrap-api.md

```diff
@@ -0,0 +1,309 @@
+| Title  | AsyncWrap API |
+|--------|---------------|
+| Author | @trevnorris   |
+| Status | DRAFT         |
+| Date   | 2016-04-05    |
+
+## Description
+
+Since its initial introduction with the `AsyncListener` API, `AsyncWrap` has
+slowly evolved to ensure a generalized API that would serve as a solid base for
+module authors who wished to add events to the life cycle of the event loop.
+The API presented below aims to accomplish this.
+
+## API
+
+### Overview
+
+```js
+// Standard way of requiring a module. Snake case follows core module
+// convention.
+const async_wrap = require('async_wrap');
+
+// List of asynchronous providers (e.g. CRYPTO or TCPWRAP). Can be used for
+// trace filtering of events.
+const providers = async_wrap.Providers;
+
+// init() is called during object construction. The object handle being
```

Comment:
My understanding of the user model:

- `init(id, type, parentId)` - invoked when an async callback as identified by `id` is eligible to be invoked.  `parentId` is the _async ID_ of the currently executing parent at the time `init()` is called.
- `pre(id)` - invoked immediately before an async callback as identified by `id` is invoked
- `post(id, hasThrown)` - invoked immediately after an async callback as identified by `id` has been invoked.
-  `destroy(id)` - invoked when an async callback as identified by `id` is no longer elligible to be called. 

Above is brief wrt the "user model".  It can be expanded on to provide more succinct definitions.  E.g., an _async callback_ is defined as a javascript function which is enqueued for execution at some later time.  It is identified by a unique ID refered to as the _async ID_.  It has the following lifecycle events:  init() - ...